src/sudoers/entry.rs

```diff
@@ -0,0 +1,193 @@
+use core::fmt;
+
+use crate::sudoers::{
+    ast::{Identifier, Qualified, UserSpecifier},
+    tokens::{ChDir, Meta},
+};
+
+use super::{
+    ast::{RunAs, Spec, Tag},
+    tokens::Command,
+};
+
+pub struct Entry<'a> {
+    run_as: &'a RunAs,
+    cmd_specs: Vec<(Tag, &'a Spec<Command>)>,
+}
+
+impl<'a> Entry<'a> {
+    pub(super) fn new(run_as: &'a RunAs, cmd_specs: Vec<(Tag, &'a Spec<Command>)>) -> Self {
+        debug_assert!(!cmd_specs.is_empty());
+
+        Self { run_as, cmd_specs }
+    }
+}
+
+impl fmt::Display for Entry<'_> {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        let Self { run_as, cmd_specs } = self;
+
+        f.write_str("    (")?;
+        write_users(run_as, f)?;
+        write_groups(run_as, f)?;
+        f.write_str(") ")?;
+
+        let mut last_tag = None;
+        for (tag, spec) in cmd_specs {
```

Comment:
I think you could avoid checking if you're in the first iteraton by creating a new iterator explicitly and returning early if it is empty
```rust
let mut cmd_specs = cmd_specs.iter();

let Some((mut last_tag, spec)) = cmd_specs.next() else { return Ok(()) };

write_tag(f, last_tag, None)?;
write_spec(f, spec)?;

for (tag, spec) in cmd_specs {
   f.write_str(", ")?;
   write_tag(f, tag, Some(last_tag))?;
   write_spec(f, spec)?;

   last_tag = tag;
}

```

Reply:
this feels less "DRY" to me since now you have one iteration loop outside the for loop so if the body the loop ever needs to be changed then you need to be careful and remember to update the code that precedes the for loop.

Reply:
hmm that's a good point. I think introducing a function for this in the future would be a way to get the "best" of both options.